lib/active_model/serializer/utils.rb

```diff
@@ -0,0 +1,42 @@
+module ActiveModel::Serializer::Utils
+  module_function
+
+    # converts the include hash to a standard format
+    # for constrained serialization recursion
+    #
+    # converts
+    #  [:author, :comments => [:author]] to
+    #  {:author => [], :comments => [:author]}
+    #
+    # and
+    #  [:author, :comments => {:author => :bio}, :posts => [:comments]] to
+    #  {:author => [], :comments => {:author => :bio}, :posts => [:comments]}
+    #
+    # The data passed in to this method should be an array where the last
+    # parameter is a hash
+    #
+    # the point of this method is to normalize the include
+    # options for the child relationships.
+    # if a sub inclusion is still an array after this method,
+    # it will get converted during the next iteration
+    def include_array_to_hash(include_array)
+      # still don't trust input
+      # but this also allows
+      #  include: :author syntax
+      include_array = Array[*include_array].compact
+
+      result = {}
+
+      hashes = include_array.select{|a| a.is_a?(Hash)}
+      non_hashes = include_array - hashes
+
+      hashes += non_hashes.map{ |association_name| { association_name => [] } }
```

Comment:
I think `reduce` is better semantically, but is a bit more verbose.  It shines when each loop returns the input for the next loop, like for summation.

```ruby
def hash_from_array(array, non_hash_value: [])
  array.reduce({}) { |hash, element| # notice the reversed block args, fun!
    if element.respond_to?(:keys)
      hash.update(element)
    else
      hash[element] = non_hash_value.dup
    end
    hash
  }
end
```